apis/core/v1beta1/openstackcontrolplane_types.go

```diff
@@ -274,10 +274,18 @@ func init() {
 
 // IsReady - returns true if service is ready to serve requests
 func (instance OpenStackControlPlane) IsReady() bool {
-	return instance.Status.Conditions.IsTrue(OpenStackControlPlaneRabbitMQReadyCondition) &&
-		instance.Status.Conditions.IsTrue(OpenStackControlPlaneMariaDBReadyCondition) &&
-		instance.Status.Conditions.IsTrue(OpenStackControlPlaneKeystoneAPIReadyCondition) &&
-		instance.Status.Conditions.IsTrue(OpenStackControlPlanePlacementAPIReadyCondition) &&
-		instance.Status.Conditions.IsTrue(OpenStackControlPlaneGlanceReadyCondition) &&
-		instance.Status.Conditions.IsTrue(OpenStackControlPlaneCinderReadyCondition)
+	return (!instance.Spec.Rabbitmq.Enabled ||
+		instance.Status.Conditions.IsTrue(OpenStackControlPlaneRabbitMQReadyCondition)) &&
+		(!instance.Spec.Mariadb.Enabled ||
+			instance.Status.Conditions.IsTrue(OpenStackControlPlaneMariaDBReadyCondition)) &&
+		(!instance.Spec.Ovn.Enabled ||
+			instance.Status.Conditions.IsTrue(OpenStackControlPlaneOVNReadyCondition)) &&
+		(!instance.Spec.Keystone.Enabled ||
+			instance.Status.Conditions.IsTrue(OpenStackControlPlaneKeystoneAPIReadyCondition)) &&
+		(!instance.Spec.Placement.Enabled ||
+			instance.Status.Conditions.IsTrue(OpenStackControlPlanePlacementAPIReadyCondition)) &&
+		(!instance.Spec.Glance.Enabled ||
+			instance.Status.Conditions.IsTrue(OpenStackControlPlaneGlanceReadyCondition)) &&
+		(!instance.Spec.Cinder.Enabled ||
+			instance.Status.Conditions.IsTrue(OpenStackControlPlaneCinderReadyCondition))
```

Comment:
What if we define this function as True if all the conditions in the list (except ReadyCondition) is true? 
As 
* the operator now removes the conditions of disabled services 
* initialize all the conditions to Unknown automatically
* all the conditions use True to report success

I think the below code has the same result with less complexity.

```golang
	for _, c := range instance.Status.Conditions {
		if c.Type == condition.ReadyCondition {
			continue
		}
		if c.Status != corev1.ConditionTrue {
			return false
		}
	}
	return true
```

Reply:
> What if we define this function as True if all the conditions in the list (except ReadyCondition) is true?

Is it not possible that the condition is not added to Status.Conditions inspite of the service being enabled? If that's not possible then we can simplify it. But I thought to keep it explicit as before. 

> initialize all the conditions to Unknown automatically
Why should we do that, unless the the service has to be reconciled?

> all the conditions use True to report success

Reply:
> I think the reasoning to do that is to let the use know what are the steps still needed to reach the Ready status.

IMO, initial reconciliation is different where the *ReadyCondition could be set to 'Unknown'. I don't think we should set all service conditions 'Unknown' for before every reconciliation. If the service is to be reconciled then only their status would change.

Reply:
Yeah it is messy. 

We have to differentiate between two cases when we say setting a service condition to Unknown.
1) Setting it temporarily at the beginning of the reconcile and letting the actual run of the Reconcile function update that Unknown state based on the current state of the service. 
2) Setting it to Unknown at the beginning *and persisting that Unknown to the API* via Update or Patch.

I agree that we should not do 2) for each Reconcile run as that will cause service conditions to flip between states rapidly. But I believe we can do 1) and let the Reconcile run do the recalculation each time and if that recalculation result in the same condition state as before persisted then the deferred Patch call will be no-op anyhow.

Reply:
Option 1 sounds reasonable

Reply:
A few additional things to consider if we do option 1.

- if you initialize status temporarily without persisting for every reconcile, what conditional  would you use for the initial feedback[1] for first reconcile (currently it checks `instance.Status.Conditions == nil`)?

- Are we going to change how we update/change status in every operator[2]?

[1] https://github.com/openstack-k8s-operators/openstack-operator/blob/master/controllers/core/openstackcontrolplane_controller.go#L170
[2] https://github.com/openstack-k8s-operators/glance-operator/blob/master/controllers/glance_controller.go#L147

Reply:
> A few additional things to consider if we do option 1.
> 
>     * if you initialize status temporarily without persisting for every reconcile, what conditional  would you use for the initial feedback[1] for first reconcile (currently it checks `instance.Status.Conditions == nil`)?

I think we can use == nil still for the initial feedback as that will be only true until the end of the first Reconcile. In the rest of the reconciles we never observer == nil and do reinit of the Conditions without returning right away and persisting it.

> 
>     * Are we going to change how we update/change status in every operator[2]?

I think this case is special as setting enable=false in openstack-operator for a service the conditions for that service becomes invalid and we want to remove that condition. While in other operator I don't see similar cases. There each condition is added at start and has a meaning throughout of the existence of the CR.

> 
> 
> [1] https://github.com/openstack-k8s-operators/openstack-operator/blob/master/controllers/core/openstackcontrolplane_controller.go#L170 [2] https://github.com/openstack-k8s-operators/glance-operator/blob/master/controllers/glance_controller.go#L147

Reply:
> I think we can use == nil still for the initial feedback as that will be only true until the end of the first Reconcile

That would make the code flow not so nice.  (check status empty, update the status and return, else update status and proceed). Anyways.

> I think this case is special as setting enable=false in openstack-operator for a service the conditions for that service becomes invalid and we want to remove that condition.

I don't see it as different. We're creating a bunch of resources (standard vs custom) and updating the status. What we're saying is for the openstackctlplane CR, status of every resource it's composed of is unknown in the beginning of reconcile. It can also apply to others too.

Reply:
I rest my case here. I think we went into a rabbit hole (sorry for that) and it is becoming unproductive. I will stop pushing for changes in the implementation itself and I will try to focus on having an acceptable external behavior. On that point my only goal would be not to have misleading information in the Status.Conditions after a service is disabled. If the current implementation has that behavior then I'm OK with it.

Reply:
@gibizer Appreciate your review and feedback. We're not going into a rabbit hole. We can disagree on certain things and still can find a compromise to move forward. My point was to have a common pattern  across operators. Also, I feel we can work and merge things iteratively. Anyway, I've updated the patch as you wanted. We can revisit it if required later.